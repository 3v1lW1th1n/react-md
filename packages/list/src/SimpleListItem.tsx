import React, { FC, forwardRef } from "react";
import cn from "classnames";
import { bem, WithForwardedRef } from "@react-md/utils";

import ListItemChildren from "./ListItemChildren";
import getListItemHeight, { SimpleListItemProps } from "./getListItemHeight";

type WithRef = WithForwardedRef<HTMLLIElement>;
type DefaultProps = Required<
  Pick<
    SimpleListItemProps,
    "height" | "clickable" | "threeLines" | "leftPosition" | "rightPosition"
  >
>;
type WithDefaultProps = SimpleListItemProps & DefaultProps & WithRef;

const block = bem("rmd-list-item");

const SimpleListItem: FC<SimpleListItemProps & WithRef> = providedProps => {
  const {
    className,
    textClassName,
    secondaryTextClassName,
    textChildren,
    primaryText,
    secondaryText,
    leftIcon,
    leftAvatar,
    leftMedia,
    leftMediaLarge,
    leftPosition,
    rightIcon,
    rightAvatar,
    rightMedia,
    rightMediaLarge,
    rightPosition,
    forceIconWrap,
    children,
    forwardedRef,
    height: propHeight,
    threeLines,
    "aria-disabled": ariaDisabled,
    disabled,
    clickable,
    ...props
  } = providedProps as WithDefaultProps;

  const height = getListItemHeight(providedProps);
  return (
    <li
      {...props}
      aria-disabled={
        typeof ariaDisabled === "string"
          ? ariaDisabled
          : (disabled && "true") || undefined
      }
      ref={forwardedRef}
      className={cn(
        block({
          [height]: height !== "auto" && height !== "normal",
          "three-lines": threeLines,
          clickable,
        }),
        className
      )}
    >
      <ListItemChildren
        textClassName={textClassName}
        secondaryTextClassName={secondaryTextClassName}
        textChildren={textChildren}
        primaryText={primaryText}
        secondaryText={secondaryText}
        leftIcon={leftIcon}
        leftAvatar={leftAvatar}
        leftMedia={leftMedia}
        leftMediaLarge={leftMediaLarge}
        leftPosition={leftPosition}
        rightIcon={rightIcon}
        rightAvatar={rightAvatar}
        rightMedia={rightMedia}
        rightMediaLarge={rightMediaLarge}
        rightPosition={rightPosition}
        forceIconWrap={forceIconWrap}
      >
        {children}
      </ListItemChildren>
    </li>
  );
};

const defaultProps: DefaultProps = {
  height: "auto",
  clickable: false,
  threeLines: false,
  leftPosition: "middle",
  rightPosition: "middle",
};

SimpleListItem.defaultProps = defaultProps;

if (process.env.NODE_ENV !== "production") {
  SimpleListItem.displayName = "SimpleListItem";

  let PropTypes = null;
  try {
    PropTypes = require("prop-types");
  } catch (e) {}

  if (PropTypes) {
    SimpleListItem.propTypes = {
      disabled: PropTypes.bool,
      clickable: PropTypes.bool,
      threeLines: PropTypes.bool,
      height: PropTypes.oneOf([
        "auto",
        "normal",
        "medium",
        "large",
        "extra-large",
      ]),
      leftPosition: PropTypes.oneOf(["top", "middle", "bottom"]),
      rightPosition: PropTypes.oneOf(["top", "middle", "bottom"]),
    };
  }
}

export default forwardRef<HTMLLIElement, SimpleListItemProps>((props, ref) => (
  <SimpleListItem forwardedRef={ref} {...props} />
));
