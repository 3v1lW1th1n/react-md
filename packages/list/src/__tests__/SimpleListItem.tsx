import React from "react";
import { render } from "@testing-library/react";

import SimpleListItem from "../SimpleListItem";

describe("SimpleListItem", () => {
  it("should render correctly", () => {
    const props = { children: "Content" };
    const { container, rerender } = render(<SimpleListItem {...props} />);

    expect(container).toMatchSnapshot();

    rerender(<SimpleListItem {...props} leftAddon={<span>Left</span>} />);
    expect(container).toMatchSnapshot();
  });
});
