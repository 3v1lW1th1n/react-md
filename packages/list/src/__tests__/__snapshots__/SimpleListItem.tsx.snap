// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SimpleListItem should render correctly 1`] = `
<div>
  <li
    class="rmd-list-item"
  >
    Content
  </li>
</div>
`;

exports[`SimpleListItem should render correctly 2`] = `
<div>
  <li
    class="rmd-list-item rmd-list-item--medium"
  >
    <span
      class="rmd-icon--before rmd-list-item__icon rmd-list-item__icon--before"
    >
      Left
    </span>
    Content
  </li>
</div>
`;
